stan/math/prim/arr/fun/matrix_gpu.hpp

```diff
@@ -25,169 +20,195 @@ enum copy_transposed_triangular {LOWER_TO_UPPER_TRIANGULAR = 0,
 namespace stan {
   namespace math {
 
+    enum triangularity {LOWER = 0, UPPER = 1, NONE = 2 };
+    enum copy_transposed_triangular {LOWER_TO_UPPER_TRIANGULAR = 0,
+      UPPER_TO_LOWER_TRIANGULAR = 1};
+
+    /**
+     * This class represents a matrix on the GPU. 
+     * 
+     * The matrix data is stored in the oclBuffer.     
+     * 
+     */
     class matrix_gpu {
       private:
-
         cl::Buffer oclBuffer;
 
       public:
+        int rows_;
+        int cols_;
+        int rows() {
+          return rows_;
+        }
 
-        int rows;
-        int cols;
-        cl::Buffer buffer() {
+        int cols() {
+          return cols_;
+        }
+
+        int size() {
+          return rows_ * cols_;
+        }
+
+        const cl::Buffer& buffer() const {
           return oclBuffer;
         }
 
-        //TODO: constructors with enumerator added when
-        // the matrix_gpu does not need to be READ_WRITE
-        matrix_gpu(int rows,  int cols) {
+        matrix_gpu() {
+          rows_ = 0;
+          cols_ = 0;
+        }
+        // TODO(Rok): constructors with enumerator added when
+        //  the matrix_gpu does not need to be READ_WRITE
+        /**
+         * Constructor for the matrix_gpu that 
+         * only allocates the buffer on the GPU.
+         * 
+         * @param rows number of matrix rows
+         * @param cols number of matrix columns
+         * 
+         * @throw <code>std::invalid_argument</code> if the
+         * matrices do not have matching dimensions
+         * 
+         */
+        matrix_gpu(int rows,  int cols)
+        : rows_(rows), cols_(cols) {
           try {
-            cl::Context ctx = stan::math::get_context();
-            oclBuffer = cl::Buffer(ctx, CL_MEM_READ_WRITE,
-             sizeof(double) * rows * cols);
-            this->rows = rows;
-            this->cols = cols;
-          } catch (cl::Error& e) {
-            check_ocl_error(e);
+            cl::Context& ctx = get_context();
+            if (rows_ > 0 && cols_ > 0) {
```

Comment:
Does this leave `oclBuffer` undefined if `rows_ > 0` or `cols_ > 0`?  Should this just throw an exception instead?  Should this check go before getting a context then?

Reply:
The stan math [tests](https://github.com/stan-dev/math/blob/b9844645a369f3187342f14abe3ded328c423583/test/unit/math/prim/mat/fun/multiply_test.cpp#L40) do not expect an exception when a matrix/vector/row_vector of size 0 is used.

The OpenCL cl::Buffer constructor throws an exception if the size of Buffer is 0, so this check was added. I moved it before the get_context, which is obviously unnecessary, if size() is 0.